constraint-solver/src/indexed_constraint_system.rs

```diff
@@ -40,18 +55,62 @@ impl<T: FieldElement, V> IndexedConstraintSystem<T, V> {
 impl<T: FieldElement, V: Clone + Hash + Ord + Eq> IndexedConstraintSystem<T, V> {
     /// Substitutes a variable with a symbolic expression in all algebraic expressions
     pub fn substitute_by_known(&mut self, variable: &V, substitution: &SymbolicExpression<T, V>) {
-        // TODO: Make this more efficient by remembering where the variable appears
-        self.constraint_system
-            .algebraic_constraints
-            .iter_mut()
-            .chain(
-                self.constraint_system
-                    .bus_interactions
-                    .iter_mut()
-                    .flat_map(|b| b.iter_mut()),
-            )
-            .for_each(|expr| {
-                expr.substitute_by_known(variable, substitution);
-            });
+        // Since we substitute by a known value, we do not need to update variable_occurrences.
+        for item in self
+            .variable_occurrences
+            .get(variable)
+            .unwrap_or(&Vec::new())
```

Comment:
Is it expected that a variable would not be in the map?

Reply:
I think this would be a bug, so seems like we can unwrap here (instead of being silent).